### PR TITLE
F 2045 2123 2124 new designed milestone forms for editing

### PR DESCRIPTION
**Pull request by RamRamez**
1. new designed milestone forms for editing
2. wrong notification after milestone update
3. Add formType field to milestone creation form and Milestone object

### CODE REVIEW

File: src/components/EditMilestoneButton.jsx [hunk 1]
```diff
@@ -20,7 +20,11 @@ const EditMilestoneButton = forwardRef(({ milestone }, ref) => {
   const goMilestoneEditPage = () => {
     checkBalance(balance)
       .then(() => {
-        if (['Proposed', 'Rejected'].includes(milestone.status)) {
+        const { formType } = milestone;
+        if (formType) {
```
Comment (aminlatifi) [inline src/components/EditMilestoneButton.jsx:24]: Limit form type possible values

File: src/containers/Routes.jsx [hunk 2]
```diff
@@ -215,6 +215,46 @@ const Routes = () => {
           />
         )}
       />
+      <Route exact path="/bounty/:milestoneId/edit" render={props => <CreateBounty {...props} />} />
+      <Route
+        exact
+        path="/expense/:milestoneId/edit"
+        render={props => (
+          <CreateExpense
+            balance={balance}
```
Comment (aminlatifi) [inline src/containers/Routes.jsx:224]: These three fields can be removed

Comment (aminlatifi) [reply src/containers/Routes.jsx:224]: Also from next two tags

File: src/components/views/CreateBounty.jsx [hunk 3]
```diff
@@ -29,17 +29,80 @@ function CreateBounty(props) {
...
+  useEffect(() => {
+    if (!isNew && currentUser.id) {
+      MilestoneService.get(milestoneId)
+        .then(res => {
+          if (
+            res.formType !== Milestone.BOUNTYTYPE ||
+            !(
+              isOwner(res.owner.address, currentUser) ||
+              isOwner(res.campaign.ownerAddress, currentUser) ||
+              res.donationCounters.length > 0
```
Comment (aminlatifi) [inline src/components/views/CreateBounty.jsx:81]: This condition is wrong, a milestone having some donations can be edited by a user who is not milestone owner neither campaign owner

File: src/components/views/CreateBounty.jsx [hunk 4]
```diff
@@ -29,17 +29,80 @@ function CreateBounty(props) {
...
+  };
+
+  useEffect(() => {
+    if (isNew) {
+      getCampaign(campaignSlug).then();
+    }
+  }, []);
+
+  useEffect(() => {
+    if (!isNew && currentUser.id) {
```
Comment (aminlatifi) [inline src/components/views/CreateBounty.jsx:73]: What if user doesn't have any `id`? a user which has not signed in or created profile

File: src/components/views/CreateBounty.jsx [hunk 5]
```diff
@@ -29,17 +29,80 @@ function CreateBounty(props) {
...
+            res.formType !== Milestone.BOUNTYTYPE ||
+            !(
+              isOwner(res.owner.address, currentUser) ||
+              isOwner(res.campaign.ownerAddress, currentUser) ||
+              res.donationCounters.length > 0
+            )
+          ) {
+            goBack();
+          } else {
+            getCampaign(res.campaign.slug).then();
```
Comment (aminlatifi) [inline src/components/views/CreateBounty.jsx:86]: Why get campaign again? milestone entity already has it

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Combining create/propose and editing forms makes views complex and hard to debug. @RamRamez Please create separate view for editing each milestone form type, like EditBounty, EditExpenses, etc.

**Comment by aminlatifi** [comment]
> It's not possible to edit title of an already accepted expense trance
> Happened on develop `/expense/609a9c9de66558424a7f9a28/edit`
> image

@RamRamez Please test it for other fields.

**Comment by aminlatifi** [comment]
@RamRamez There is an issue on setting conversion rates
```
index.js:1 save milestone error 
BadRequest {type: "FeathersError", name: "BadRequest", message: "Validation failed: items.0.conversionRateTimestamp…tedFiatType: Path `selectedFiatType` is required.", code: 400, className: "bad-request", …}
className: "bad-request"
code: 400
data: undefined
errors: {items.0.conversionRateTimestamp: {…}, items.0.conversionRate: {…}, items.0.wei: {…}, items.0.fiatAmount: {…}, items.0.selectedFiatType: {…}, …}
hook: {type: "before", arguments: Array(2), service: {…}, app: {…}, method: "patch", …}
message: "Validation failed: items.0.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.0.conversionRate: Path `conversionRate` is required., items.0.wei: Path `wei` is required., items.0.fiatAmount: Path `fiatAmount` is required., items.0.selectedFiatType: Path `selectedFiatType` is required., items.1.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.1.conversionRate: Path `conversionRate` is required., items.1.wei: Path `wei` is required., items.1.fiatAmount: Path `fiatAmount` is required., items.1.selectedFiatType: Path `selectedFiatType` is required."
name: "BadRequest"
type: "FeathersError"
stack: "BadRequest: Validation failed: items.0.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.0.conversionRate: Path `conversionRate` is required., items.0.wei: Path `wei` is required., items.0.fiatAmount: Path `fiatAmount` is required., items.0.selectedFiatType: Path `selectedFiatType` is required., items.1.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.1.conversionRate: Path `conversionRate` is required., items.1.wei: Path `wei` is required., items.1.fiatAmount: Path `fiatAmount` is required., items.1.selectedFiatType: Path `selectedFiatType` is required.\n    at new BadRequest (http://localhost:3010/static/js/vendors~main.chunk.js:178438:17)\n    at convert (http://localhost:3010/static/js/vendors~main.chunk.js:178574:32)\n    at Socket.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:182484:47)\n    at Socket.push../node_modules/socket.io-client/lib/socket.js.Socket.onack (http://localhost:3010/static/js/vendors~main.chunk.js:393322:9)\n    at Socket.push../node_modules/socket.io-client/lib/socket.js.Socket.onpacket (http://localhost:3010/static/js/vendors~main.chunk.js:393247:12)\n    at Manager.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Manager.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Manager.push../node_modules/socket.io-client/lib/manager.js.Manager.ondecoded (http://localhost:3010/static/js/vendors~main.chunk.js:392741:8)\n    at Decoder.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Decoder.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Decoder.push../node_modules/socket.io-parser/index.js.Decoder.add (http://localhost:3010/static/js/vendors~main.chunk.js:394508:12)\n    at Manager.push../node_modules/socket.io-client/lib/manager.js.Manager.ondata (http://localhost:3010/static/js/vendors~main.chunk.js:392731:16)\n    at Socket.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Socket.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Socket.push../node_modules/engine.io-client/lib/socket.js.Socket.onPacket (http://localhost:3010/static/js/vendors~main.chunk.js:166181:14)\n    at WS.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:166007:10)\n    at WS.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at WS.push../node_modules/engine.io-client/lib/transport.js.Transport.onPacket (http://localhost:3010/static/js/vendors~main.chunk.js:166610:8)\n    at WS.push../node_modules/engine.io-client/lib/transport.js.Transport.onData (http://localhost:3010/static/js/vendors~main.chunk.js:166602:8)\n    at WebSocket.ws.onmessage (http://localhost:3010/static/js/vendors~main.chunk.js:167781:10)"
__proto__: Error
```

**Comment by RamRamez** [comment]
> @RamRamez There is an issue on setting conversion rates
> 
> ```
> index.js:1 save milestone error 
> BadRequest {type: "FeathersError", name: "BadRequest", message: "Validation failed: items.0.conversionRateTimestamp…tedFiatType: Path `selectedFiatType` is required.", code: 400, className: "bad-request", …}
> className: "bad-request"
> code: 400
> data: undefined
> errors: {items.0.conversionRateTimestamp: {…}, items.0.conversionRate: {…}, items.0.wei: {…}, items.0.fiatAmount: {…}, items.0.selectedFiatType: {…}, …}
> hook: {type: "before", arguments: Array(2), service: {…}, app: {…}, method: "patch", …}
> message: "Validation failed: items.0.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.0.conversionRate: Path `conversionRate` is required., items.0.wei: Path `wei` is required., items.0.fiatAmount: Path `fiatAmount` is required., items.0.selectedFiatType: Path `selectedFiatType` is required., items.1.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.1.conversionRate: Path `conversionRate` is required., items.1.wei: Path `wei` is required., items.1.fiatAmount: Path `fiatAmount` is required., items.1.selectedFiatType: Path `selectedFiatType` is required."
> name: "BadRequest"
> type: "FeathersError"
> stack: "BadRequest: Validation failed: items.0.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.0.conversionRate: Path `conversionRate` is required., items.0.wei: Path `wei` is required., items.0.fiatAmount: Path `fiatAmount` is required., items.0.selectedFiatType: Path `selectedFiatType` is required., items.1.conversionRateTimestamp: Path `conversionRateTimestamp` is required., items.1.conversionRate: Path `conversionRate` is required., items.1.wei: Path `wei` is required., items.1.fiatAmount: Path `fiatAmount` is required., items.1.selectedFiatType: Path `selectedFiatType` is required.\n    at new BadRequest (http://localhost:3010/static/js/vendors~main.chunk.js:178438:17)\n    at convert (http://localhost:3010/static/js/vendors~main.chunk.js:178574:32)\n    at Socket.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:182484:47)\n    at Socket.push../node_modules/socket.io-client/lib/socket.js.Socket.onack (http://localhost:3010/static/js/vendors~main.chunk.js:393322:9)\n    at Socket.push../node_modules/socket.io-client/lib/socket.js.Socket.onpacket (http://localhost:3010/static/js/vendors~main.chunk.js:393247:12)\n    at Manager.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Manager.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Manager.push../node_modules/socket.io-client/lib/manager.js.Manager.ondecoded (http://localhost:3010/static/js/vendors~main.chunk.js:392741:8)\n    at Decoder.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Decoder.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Decoder.push../node_modules/socket.io-parser/index.js.Decoder.add (http://localhost:3010/static/js/vendors~main.chunk.js:394508:12)\n    at Manager.push../node_modules/socket.io-client/lib/manager.js.Manager.ondata (http://localhost:3010/static/js/vendors~main.chunk.js:392731:16)\n    at Socket.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:148671:15)\n    at Socket.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at Socket.push../node_modules/engine.io-client/lib/socket.js.Socket.onPacket (http://localhost:3010/static/js/vendors~main.chunk.js:166181:14)\n    at WS.<anonymous> (http://localhost:3010/static/js/vendors~main.chunk.js:166007:10)\n    at WS.push../node_modules/component-emitter/index.js.Emitter.emit (http://localhost:3010/static/js/vendors~main.chunk.js:148823:20)\n    at WS.push../node_modules/engine.io-client/lib/transport.js.Transport.onPacket (http://localhost:3010/static/js/vendors~main.chunk.js:166610:8)\n    at WS.push../node_modules/engine.io-client/lib/transport.js.Transport.onData (http://localhost:3010/static/js/vendors~main.chunk.js:166602:8)\n    at WebSocket.ws.onmessage (http://localhost:3010/static/js/vendors~main.chunk.js:167781:10)"
> __proto__: Error
> ```

Thanks, t's fixed.

**Review by aminlatifi — COMMENTED** [review]
I created a Payment trace (milestone) with capacity. Then I edited it and made it uncapped (check no limit). It didn't work.

File: src/components/views/EditMilestone.jsx [hunk 6]
```diff
@@ -1,1405 +1,324 @@
...
+    const ms = milestone;
+    if (type === 'checkbox' && name === 'hasReviewer') {
+      setHasReviewer(checked);
+    } else if (type === 'checkbox' && name === 'donateToDac') {
+      setDonateToDac(checked);
+    } else if (name === 'image') {
+      setImage(value);
+    } else {
+      ms[name] = value;
+      setMilestone(ms);
```
Comment (aminlatifi) [inline src/components/views/EditMilestone.jsx:128]: This probably has no effect

File: src/components/views/EditPayment.jsx [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+      milestone.parentProjectId = campaign.projectId;
+      milestone.title = title;
+      milestone.description = description;
+      milestone.recipientAddress = recipientAddress;
+      milestone.image = image;
+      milestone.token = token;
+      milestone.dacId = donateToDac ? config.defaultDacId : 0;
+      // TODO: We should have ability to delete fiatAmount for uncapped milestones
+
+      if (!notCapped) {
```
Comment (aminlatifi) [inline src/components/views/EditPayment.jsx:259]: We should have an else statement for this condition and clear max amount and other fields if milestone is not capped.

@RamRamez refer to this
https://github.com/Giveth/giveth-dapp/pull/2126#pullrequestreview-660389475

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Thank you @RamRamez . Please resolve mentioned comments and let's go for test 😃

**Comment by aminlatifi** [comment]
Thanks @RamRamez

**Pull request merged**